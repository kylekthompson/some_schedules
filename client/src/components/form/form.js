import PropTypes from 'prop-types';
import React, { Component } from 'react';
import produce from 'immer';
import styled from 'styled-components';
import { memoize } from 'helpers/function';

const StyledForm = styled.form`
  display: flex;
  flex-direction: column;
`;

export default class Form extends Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    className: PropTypes.string,
    fields: PropTypes.arrayOf(PropTypes.shape({
      initialValue: PropTypes.any,
      name: PropTypes.string.isRequired,
    })).isRequired,
    onSubmit: PropTypes.func,
  };

  static defaultProps = {
    onSubmit: () => {},
  };

  state = this.props.fields.reduce((state, field) => ({
    ...state,
    [field.name]: {
      didBlur: false,
      errors: [],
      isDirty: false,
      value: field.initialValue || '',
    },
  }), {});

  get renderProps() {
    return {
      errorPropsForField: this.errorPropsForField,
      inputPropsForField: this.inputPropsForField,
      submitProps: this.submitProps,
    };
  }

  get submitProps() {
    return {
      disabled: this.isSubmitDisabled,
    };
  }

  get isSubmitDisabled() {
    return this.props.fields.some(({ name }) => this.state[name].errors.length > 0);
  }

  errorPropsForField = (field) => ({
    errors: this.formErrors()[field],
  });

  inputPropsForField = (field) => ({
    onBlur: this.handleBlur(field),
    onChange: this.handleChange(field),
    value: this.formValues()[field],
  });

  pluckFormField = (formField) => (state = this.state) => this.props.fields.reduce((values, field) => ({
    ...values,
    [field.name]: state[field.name][formField],
  }), {});
  formValues = this.pluckFormField('value');
  formErrors = this.pluckFormField('errors');

  handleBlur = memoize((field) => (event) => {
    const { value } = event.currentTarget;

    this.setState(produce((state) => {
      const isDirty = state[field].didBlur || Boolean(value);
      let errors = [];

      if (isDirty) {
        errors = this.props.fields.find(({ name }) => name === field).validations.run({
          ...this.formValues(state),
          [field]: value,
        });
      }

      state[field].didBlur = isDirty;
      state[field].errors = errors;
      state[field].isDirty = true;
      state[field].value = value;
    }));
  });

  handleChange = memoize((field) => (event) => {
    const { value } = event.currentTarget;

    this.setState(produce((state) => {
      let errors = [];

      if (state[field].didBlur && state[field].isDirty) {
        errors = this.props.fields.find(({ name }) => name === field).validations.run({
          ...this.formValues(state),
          [field]: value,
        });
      }

      state[field].errors = errors;
      state[field].isDirty = true;
      state[field].value = value;
    }));
  });

  forceValidation = (callback) => {
    this.setState(produce((state) => {
      this.props.fields.forEach((field) => {
        state[field.name].didBlur = true;
        state[field.name].errors = field.validations.run(this.formValues(state));
        state[field.name].isDirty = true;
      });
    }), callback);
  }

  handleSubmit = (event) => {
    event.preventDefault();

    this.forceValidation(() => {
      if (this.isSubmitDisabled) {
        return;
      }

      this.props.onSubmit(this.formValues());
    });
  };

  render() {
    return (
      <StyledForm className={this.props.className} onSubmit={this.handleSubmit}>
        {this.props.children(this.renderProps)}
      </StyledForm>
    );
  }
}
