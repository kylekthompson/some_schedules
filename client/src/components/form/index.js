export { default as Errors } from 'components/form/errors';
export { default as Form } from 'components/form/form';
export { default as Input } from 'components/form/input';
export { default as Separator } from 'components/form/separator';
