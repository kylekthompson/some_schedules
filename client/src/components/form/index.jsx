import Errors from 'components/form/errors';
import {
  Container,
  Form,
  HeaderContainer,
  Input,
  Label,
  Subtitle,
  Submit,
  Title,
} from 'components/form/styled-components';

Form.Container = Container;
Form.Errors = Errors;
Form.HeaderContainer = HeaderContainer;
Form.Input = Input;
Form.Label = Label;
Form.Submit = Submit;
Form.Subtitle = Subtitle;
Form.Title = Title;

export default Form;
