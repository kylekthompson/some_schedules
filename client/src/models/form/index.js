export { default as formValuesFromState } from 'models/form/formValuesFromState';
