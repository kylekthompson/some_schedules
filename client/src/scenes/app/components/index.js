export { default as Container } from 'scenes/App/components/Container';
export { default as HeaderContainer } from 'scenes/App/components/HeaderContainer';
export { default as HeaderLinks } from 'scenes/App/components/HeaderLinks';
