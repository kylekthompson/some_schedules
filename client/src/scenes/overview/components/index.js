export { default as Container } from 'scenes/Overview/components/Container';
export { default as HeaderLinks } from 'scenes/Overview/components/HeaderLinks';
export { default as Routes } from 'scenes/Overview/components/Routes';
