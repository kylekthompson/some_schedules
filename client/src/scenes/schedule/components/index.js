export { default as Container } from 'scenes/Schedule/components/Container';
export { default as ContentContainer } from 'scenes/Schedule/components/ContentContainer';
export { default as SidebarContainer } from 'scenes/Schedule/components/SidebarContainer';
