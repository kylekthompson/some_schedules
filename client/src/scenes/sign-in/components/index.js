export { default as Container } from 'scenes/sign-in/components/container';
