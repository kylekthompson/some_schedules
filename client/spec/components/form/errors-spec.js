import React from 'react';
import { Errors } from 'components/form';
import { mount } from 'spec/support/mount';

describe('<Errors />', () => {
  describe('when there are no errors', () => {
    it('renders nothing', () => {
      const { container } = mount(<Errors errors={[]} id="errors" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('when there are errors', () => {
    it('renders all of them', () => {
      const { container } = mount(<Errors errors={['Some error.', 'Some other error.']} id="errors" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
