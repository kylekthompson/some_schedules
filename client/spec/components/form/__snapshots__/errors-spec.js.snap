// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Errors /> when there are errors renders all of them 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  margin-bottom: 10px;
}

.c2 {
  font-family: 'Open Sans',sans-serif;
  font-weight: 400;
  font-size: 12px;
  color: rgba(244,96,54,0.8);
}

<div
  aria-live="polite"
  class="c0 c1"
  id="errors"
  role="alert"
>
  <p
    class="c2"
    color="rgba(244, 96, 54, 0.8)"
  >
    Some error.
  </p>
  <p
    class="c2"
    color="rgba(244, 96, 54, 0.8)"
  >
    Some other error.
  </p>
</div>
`;

exports[`<Errors /> when there are no errors renders nothing 1`] = `null`;
